### PR TITLE
implement proficiencies being used for books

### LINKED ISSUE #44680: Assisting with proficiencies through books
# Is your feature request related to a problem? Please describe.
https://github.com/CleverRaven/Cataclysm-DDA/issues/31610 is nearly complete to a point where, although there are a lot more proficiencies to add, they are at least a big and growing part of the game.

Unfortunately the only way to learn proficiencies currently is through practice. Recipe books can tell you things to make, but don't grant you any help with actually making them.

# Describe the solution you'd like
Allow books to provide partial alleviation of proficiency penalties.

A json definition for this would look like:
```{
  "id": "book_lockpick",
  "type": "BOOK",
  "name": { "str": "MIT Guide to Lock Picking",  "str_pl": "copies of MIT Guide to Lock Picking" },
  "description": "A home-made booklet with large black-and-white pictures of several types of modern locks and general information on how to crack them open.",
  "proficiencies": [ { "proficiency": "prof_lockpicking", "time_factor": 0.9, "fail_factor": 0.5, "include_prereqs": true }  ]
```
In this example, having `book_lockpick` within range would let you do tasks that require prof_lockpicking. Any tasks that have a `time_multiplier` due to their proficiency requirements would have that time multiplier reduced to 90% of its normal value, and any tasks with a `fail_multiplier` due to their proficiency requirements would have that fail multiplier reduced by half. In this example the book doesn't really speed you up, but does make you more likely to succeed.

`include_prereqs` means that the book grants this proficiency *and any other necessary prior prerequisite proficiencies for the task*, with the same multipliers to those proficiencies. For example, a blacksmithing book with this set to true should confer metalworking at the same factor as it confers blacksmithing.

If not specified, both of these values should default to `0.5`. `include_prereqs` should default to true.

## Multiple proficiency books
If you have multiple books granting a proficiency, the time factors shouldn't quite stack. We can use sum of squares for this, like we do with morale, but it has to be tweaked a little since in this case lower is better.

`Final_time_factor = 1 - sqrt( SUM[ 1 - time_factor]^2 )`

In this example having 3 books with a time factor of 0.5 will bring your time factor down to 0.36. Having three books with a fail factor of 0.9 will bring your fail factor down to 0.82. This keeps you from stacking a bunch of low-quality references, but lets you use a bunch of low quality references if that is all you have.

# Describe alternatives you've considered
It might be worth considering having the intelligence requirement of the book affect whether or not you can use it to learn proficiencies. If you have lower than the required intelligence, the `time` and `fail` factors could be multiplied by `(required int) / (actual int)` to a maximum of 1.

Likewise for required skill.

# Additional context
Once we have the ability to reduce proficiency penalties by a factor, the same infrastructure could and should be used to make proficiency penalties gradually fall off as you get close to the necessary xp to have the proficiency.

Additionally, it would be nice if having a faction NPC with a proficiency help would count as having that proficiency yourself. I don't know if we have that code already.

I think we need at least a rudimentary version of this implemented for 0.F given the number of proficiencies going in. However I have tried to set up proficiencies to work without this feature, so it's a matter of some debate.

### PR DESCRIPTION
**Pull request by KorGgenT**
#### Summary
```SUMMARY: Features "implement proficiencies being used for books"```

#### Purpose of change
Resolves #44680

#### Describe the solution
as requested by @I-am-Erk , proficiency fail and time modifiers will now also be modified by the books available to you in your crafting inventory. this adds a struct to hold the data, loading that data into a vector in an islot_book, a new class as a container for the vector with methods to do all the calculations behind the scenes, and methods in inventory in order to use crafting_inventory() easily.

#### Testing
still needs json and also testing. it compiles.

### CODE REVIEW

**Comment by I-am-Erk** [comment]
We'll also need to display proficiency bonus on the book item

**Comment by I-am-Erk** [comment]
image
Ignoring that:
image

UI stuff needs some work, which could be in a different PR, but it's working. Although I got the numbers wrong in my books.

**Comment by I-am-Erk** [comment]
So I think the only thing this needs to merge is a UI line on the book items themselves saying what proficiency bonunses they offer, and possibly for the crafting menu to mention either what the bonuses from books are, or at least that there are bonuses from books being applied.

File: data/json/items/book/fabrication.json [hunk 1]
```diff
@@ -371,6 +386,7 @@
     "skill": "fabrication",
     "max_level": 2,
     "intelligence": 5,
+    "proficiencies": [ { "proficiency": "prof_carpentry_basic", "time_factor": 0.4, "fail_factor": 0.8, "include_prereqs": true }, { "proficiency": "prof_plumbing", "time_factor": 0.5, "fail_factor": 0.8, "include_prereqs": true } ],
```
Comment (actual-nh) [inline data/json/items/book/fabrication.json:389]:
```suggestion
    "proficiencies": [
      { "proficiency": "prof_carpentry_basic", "time_factor": 0.4, "fail_factor": 0.8, "include_prereqs": true },
      { "proficiency": "prof_plumbing", "time_factor": 0.5, "fail_factor": 0.8, "include_prereqs": true }
    ],
```

File: data/json/items/book/fabrication.json [hunk 2]
```diff
@@ -389,6 +405,7 @@
     "skill": "fabrication",
     "required_level": 5,
     "max_level": 8,
+    "proficiencies": [ { "proficiency": "prof_carpentry_basic", "time_factor": 0.7, "fail_factor": 0.3, "include_prereqs": true }, { "proficiency": "prof_plumbing", "time_factor": 0.85, "fail_factor": 0.5, "include_prereqs": true } ],
```
Comment (actual-nh) [inline data/json/items/book/fabrication.json:408]:
```suggestion
    "proficiencies": [
      { "proficiency": "prof_carpentry_basic", "time_factor": 0.7, "fail_factor": 0.3, "include_prereqs": true },
      { "proficiency": "prof_plumbing", "time_factor": 0.85, "fail_factor": 0.5, "include_prereqs": true }
    ],
```

File: data/json/items/book/fabrication.json [hunk 3]
```diff
@@ -425,6 +443,7 @@
     "color": "green",
     "skill": "fabrication",
     "max_level": 3,
+    "proficiencies": [ { "proficiency": "prof_carpentry_basic", "time_factor": 0.5, "fail_factor": 0.4, "include_prereqs": true }, { "proficiency": "prof_plumbing", "time_factor": 0.5, "fail_factor": 0.6, "include_prereqs": true } ],
```
Comment (actual-nh) [inline data/json/items/book/fabrication.json:446]:
```suggestion
    "proficiencies": [
      { "proficiency": "prof_carpentry_basic", "time_factor": 0.5, "fail_factor": 0.4, "include_prereqs": true },
      { "proficiency": "prof_plumbing", "time_factor": 0.5, "fail_factor": 0.6, "include_prereqs": true }
    ],
```

**Review by actual-nh — COMMENTED** [review]
I am unsure if the ranged_balance thing is a problem, or just a false positive on a statistical test. See #46256.

**Pull request closed without merging**